client/components/phone-input/index.jsx

```diff
@@ -109,13 +111,39 @@ const PhoneInput = React.createClass( {
 		event.preventDefault();
 
 		const country = this.guessCountryFromValue( value );
-		this.props.onChange( { value, countryCode: country.isoCode } );
+		if ( this.state.freezeSelection && value.indexOf( country.dialCode ) === -1 ) {
+			return;
+		}
+		if ( country ) {
+			// todo ??
+			this.props.onChange( {
+				value: this.format( { value, countryCode: country.isoCode } ),
+				countryCode: country.isoCode
+			} );
+		} else if ( value.length > 3 ) {
+			this.props.onChange( {
+				value: value,
+				countryCode: 'world'
+			} );
+		} else {
+			this.props.onChange( {
+				value: value,
+				countryCode: this.props.countryCode
+			} );
+		}
 	},
 
 	handleCountrySelection( event ) {
 		const countryCode = event.target.value.toLowerCase();
-
-		this.props.onChange( { countryCode, value: this.props.value } );
+		let value = this.props.value;
+		// if the country changes, we fix the dial code
+		if ( countryCode !== this.props.countryCode ) {
+			value = this.props.value.replace(
+				this.getCountry( this.props.countryCode ).dialCode,
+				this.getCountry( countryCode ).dialCode
+			);
+		}
+		this.props.onChange( { countryCode, value: this.format( { value, countryCode } ) } );
```

Comment:
I think we should only trigger this callback when there was an actual change - so it should go inside the `if` just above.